tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -38,131 +40,161 @@ import           Source.Span
-            end = fromIntegral (nodeEndByte node)
-        pure (Range start end)
-      Nothing -> fail "expected a node but didn't get one"
-
-instance Unmarshal Span where
-  unmarshalNodes _ = do
-    node <- peekNode
-    case node of
-      Just node -> do
-        let spanStart = pointToPos (nodeStartPoint node)
-            spanEnd = pointToPos (nodeEndPoint node)
-        pure (Span spanStart spanEnd)
-      Nothing -> fail "expected a node but didn't get one"
+instance (UnmarshalAnn a, UnmarshalAnn b) => UnmarshalAnn (a,b) where
+  unmarshalAnn node = (,)
+    <$> unmarshalAnn @a node
+    <*> unmarshalAnn @b node
+
+instance UnmarshalAnn Range where
+  unmarshalAnn node = do
+    let start = fromIntegral (nodeStartByte node)
+        end   = fromIntegral (nodeEndByte node)
+    pure (Range start end)
+
+instance UnmarshalAnn Span where
+  unmarshalAnn node = do
+    let spanStart = pointToPos (nodeStartPoint node)
+        spanEnd   = pointToPos (nodeEndPoint node)
+    pure (Span spanStart spanEnd)
 
 pointToPos :: TSPoint -> Pos
 pointToPos (TSPoint line column) = Pos (fromIntegral line) (fromIntegral column)
 
-instance Unmarshal a => Unmarshal (Maybe a) where
-  unmarshalNodes [] = pure Nothing
-  unmarshalNodes listOfNodes = Just <$> unmarshalNodes listOfNodes
-
-instance (Unmarshal (f a), Unmarshal (g a), SymbolMatching (f a), SymbolMatching (g a)) => Unmarshal ((f :+: g) a) where
-  unmarshalNodes [node] = do
-    let lhsSymbolMatch = symbolMatch (Proxy @(f a)) node
-        rhsSymbolMatch = symbolMatch (Proxy @(g a)) node
-    if lhsSymbolMatch
-      then L1 <$> unmarshalNodes @(f a) [node]
-      else if rhsSymbolMatch
-        then R1 <$> unmarshalNodes @(g a) [node]
-        else fail $ showFailure (Proxy @((f :+: g) a)) node
-  unmarshalNodes [] = fail "expected a node of type ((f :+: g) a) but didn't get one"
-  unmarshalNodes _ = fail "expected a node of type ((f :+: g) a) but got multiple"
-
-
-instance Unmarshal a => Unmarshal [a] where
-  unmarshalNodes (x:xs) = do
-    head' <- unmarshalNodes [x]
-    tail' <- unmarshalNodes xs
-    pure $ head' : tail'
-  unmarshalNodes [] = pure []
 
+-- | Optional/repeated fields occurring in product datatypes are wrapped in type constructors, e.g. 'Maybe', '[]', or 'NonEmpty', and thus can unmarshal zero or more nodes for the same field name.
+class UnmarshalField t where
+  unmarshalField
+    :: ( Carrier sig m
+       , Member (Reader ByteString) sig
+       , Member (Reader (Ptr Cursor)) sig
+       , MonadFail m
+       , MonadIO m
+       , Unmarshal f
+       , UnmarshalAnn a
+       )
+    => [Node]
+    -> m (t (f a))
+
+instance UnmarshalField Maybe where
+  unmarshalField []  = pure Nothing
+  unmarshalField [x] = Just <$> unmarshalNode x
+  unmarshalField _   = fail "expected a node of type ((f :+: g) a) but got multiple"
+
+instance UnmarshalField [] where
+  unmarshalField (x:xs) = do
+    head' <- unmarshalNode x
+    tail' <- unmarshalField xs
```

Comment:
The control flow here is a lot more obvious than it was before: we unmarshal each node in the field to an element in the list.